File: code/algo/getoffer/go/2/lru_cache.go
package main

import "container/list"

type LRUCache struct {
	capacity int
	cache    map[int]*list.Element
	list     *list.List
}

type entry struct {
	key   int
	value int
}

func NewLRUCache(capacity int) LRUCache {
	return LRUCache{
		capacity: capacity,
		cache:    make(map[int]*list.Element),
		list:     list.New(),
	}
}

func (this *LRUCache) Get(key int) int {
	if elem, ok := this.cache[key]; ok {
		this.list.MoveToFront(elem)
		return elem.Value.(*entry).value
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if elem, ok := this.cache[key]; ok {
		elem.Value.(*entry).value = value
		this.list.MoveToFront(elem)
		return
	}
	if this.list.Len() == this.capacity {
		lastElem := this.list.Back()
		delete(this.cache, lastElem.Value.(*entry).key)
		this.list.Remove(lastElem)
	}
	newElem := this.list.PushFront(&entry{key: key, value: value})
	this.cache[key] = newElem
}

File: code/algo/getoffer/c++/2/lru_cache.cpp
#include <list>
#include <unordered_map>

class LRUCache {
public:
  LRUCache(int capacity) : capacity(capacity) {}

  int get(int key) {
    auto it = cache.find(key);
    if (it == cache.end()) {
      return -1;
    }
    // Move the accessed element to the front of the list
    list.splice(list.begin(), list, it->second);
    return it->second->second;
  }

  void put(int key, int value) {
    auto it = cache.find(key);
    if (it != cache.end()) {
      // Update the value and move the element to the front of the list
      it->second->second = value;
      list.splice(list.begin(), list, it->second);
      return;
    }
    if (list.size() == capacity) {
      // Remove the least recently used element
      int lruKey = list.back().first;
      list.pop_back();
      cache.erase(lruKey);
    }
    // Insert the new element at the front of the list
    list.emplace_front(key, value);
    cache[key] = list.begin();
  }

private:
  int capacity;
  std::list<std::pair<int, int>> list;
  std::unordered_map<int, std::list<std::pair<int, int>>::iterator> cache;
};
